src/RippleKey.cpp

```diff
@@ -0,0 +1,199 @@
+    return RippleKey{};
+}
+
+RippleKey
+RippleKey::make_RippleKey(
+    boost::filesystem::path const& keyFile)
+{
+    using namespace ripple;
+
+    std::ifstream ifsKeys (keyFile.string (), std::ios::in);
+
+    if (! ifsKeys)
+        throw std::runtime_error (
+            "Failed to open key file: " + keyFile.string());
+
+    Json::Reader reader;
+    Json::Value jKeys;
+    if (! reader.parse (ifsKeys, jKeys))
+    {
+        throw std::runtime_error (
+            "Unable to parse json key file: " + keyFile.string());
+    }
+
+    static std::array<Json::StaticString, 2> const requiredFields {{
+        jss::key_type,
+        jss::master_seed
+    }};
+
+    for (auto field : requiredFields)
+    {
+        if (! jKeys.isMember(field))
+        {
+            throw std::runtime_error (
+                std::string{ "Field '" } + field.c_str() +
+                "' is missing from key file: " +
+                keyFile.string());
+        }
+    }
+
+    auto const keyType = keyTypeFromString (jKeys[jss::key_type].asString());
+    if (keyType == KeyType::invalid)
+    {
+        throw std::runtime_error (
+            "Invalid 'key_type' field \"" +
+            jKeys[jss::key_type].asString() +
+            "\" found in key file: " +
+            keyFile.string());
+    }
+
+    return RippleKey::make_RippleKey (
+        keyType, jKeys[jss::master_seed].asString());
+}
+
+void
+RippleKey::writeToFile(boost::filesystem::path const& keyFile) const
+{
+    using namespace ripple;
+    using namespace boost::filesystem;
+
+    Json::Value jv(Json::objectValue);
+    jv[jss::key_type] = to_string(keyType_);
+    jv[jss::master_seed] = toBase58(seed_);
+    jv[jss::master_seed_hex] = strHex(seed_.data(), seed_.size());
+    jv[jss::master_key] = seedAs1751(seed_);
+    jv[jss::account_id] = toBase58(calcAccountID(publicKey_));
+    jv[jss::public_key] = toBase58(TOKEN_ACCOUNT_PUBLIC, publicKey_);
+    jv[jss::public_key_hex] = strHex(publicKey_.data(), publicKey_.size());
+    jv["secret_key"] = toBase58(TOKEN_ACCOUNT_SECRET, secretKey_);
+    jv["secret_key_hex"] = strHex(secretKey_.data(), secretKey_.size());
+
+    if (!keyFile.parent_path().empty())
+    {
+        boost::system::error_code ec;
+        if (!exists(keyFile.parent_path()))
+            boost::filesystem::create_directories(keyFile.parent_path(), ec);
+
+        if (ec || !is_directory(keyFile.parent_path()))
+            throw std::runtime_error("Cannot create directory: " +
+                keyFile.parent_path().string());
+    }
+
+    std::ofstream o(keyFile.string(), std::ios::trunc);
+    if (o.fail())
+        throw std::runtime_error("Cannot open key file: " +
+            keyFile.string());
+
+    o << jv.toStyledString();
+}
+
+void
+RippleKey::singleSign(ripple::STTx& tx) const
+{
+    using namespace ripple;
+    tx.setFieldVL(sfSigningPubKey, publicKey_.slice());
+    tx.makeFieldAbsent(sfSigners);
+    tx.sign(publicKey_, secretKey_);
```

Comment:
I still haven't caught up yet.  I believe it's a required field in the transaction that gets sent to the server.  Putting it in transaction allows the signature can be checked before we try to apply the transaction to the ledger.  So, the way I read it, this code is setting the required field.  I'm still missing something...